src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -128,7 +130,14 @@ public override async Task<(IReadOnlyList<PackageSpec> dgSpecs, IReadOnlyList<IA
                     throw new InvalidOperationException(
                         string.Format(Strings.ProjectNotLoaded_RestoreFailed, ProjectName));
                 }
-                context?.PackageSpecCache.Add(_projectFullPath, packageSpec);
+
+                if (context != null)
+                {
+                    lock (context.PackageSpecCache)
```

Comment:
Is this intentional? As per [doc](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/lock-statement#guidelines), `lock on a dedicated object instance - for example,` 
```cs 
private readonly object balanceLock = new object();
```
`or another instance that is unlikely to be used as a lock object by unrelated parts of the code`

Reply:
@kartheekp-ms  It was not intentional. I need to use actual `lock` as you suggested. 
But actual problem here `context.PackageSpecCach` already public API. So I can't prevent someone directly add value without using lock in the future. 
Dictionaries in `DependencyGraphCacheContext` need to change, but it's maybe too much change without much gain.